src/main/kotlin/no/item/xp/plugin/GenerateTypeScriptTask.kt

```diff
@@ -1,27 +1,76 @@
 package no.item.xp.plugin
 
 import java.io.File
+import java.nio.file.Path
+import java.nio.file.Paths
 import javax.inject.Inject
+import javax.xml.parsers.DocumentBuilderFactory
+import no.item.xp.plugin.models.GeneratedField
+import no.item.xp.plugin.models.XmlFile
+import no.item.xp.plugin.models.XmlType
+import no.item.xp.plugin.parser.Parser
+import no.item.xp.plugin.util.GenerateInterfaceName
+import no.item.xp.plugin.writer.TypeScriptWriter
 import org.gradle.api.DefaultTask
 import org.gradle.api.tasks.TaskAction
+import org.w3c.dom.Document
 
 open class GenerateTypeScriptTask @Inject constructor(private val extension: GenerateTypeScriptExtension) : DefaultTask() {
 
     @TaskAction
     fun generateTypeScript() {
       val files = findXmlFiles()
-      println("Found XML files: ")
-      println("--------------------")
-      files.forEach { println(it) }
-      // 2.send dem til en parser
-      // 3.send resultatet til en writer
-      // 4.exit
-    }
-
+      files.forEach { parseFile(it) }
+  }
   private fun findXmlFiles(): Sequence<String> {
     return File(extension.fileDir)
       .walk()
-      .filter { it.isFile && it.extension == "xml" }
+      .filter { it.isFile }
       .map { it.absolutePath }
   }
+
+  private fun parseFile(it: String) {
+    val xmlType: XmlType? = getXmlType(it)
+    val xmlFile: XmlFile? = xmlType?.let { it1 -> generateXmlFile(it, it1) }
+    if (xmlFile != null) {
+      val document: Document? = readXml(xmlFile)
+      val instance: Parser = Parser.create()
+      instance.parse(document)
+
+      val interfaceName: String? = GenerateInterfaceName(File(it))
+      val xml: List<GeneratedField> = instance.parse(document)
+      if (interfaceName != null) {
+        TypeScriptWriter(interfaceName, xmlType, xml)
+      }
+    }
+  }
+
+  private fun getXmlType(it: String): XmlType? {
+    val child: Path = Paths.get(it).toAbsolutePath()
+    if (isChild(child, "../site/content-types/")) { return XmlType.CONTENT_TYPE }
+    if (isChild(child, "../site/layouts/")) { return XmlType.LAYOUT }
+    if (isChild(child, "../site/mixins/")) { return XmlType.MIXIN }
+    if (isChild(child, "../site/parts/")) { return XmlType.PART }
+    if (isChild(child, "../site/pages/")) { return XmlType.PAGE }
+    if (isChild(child, "../site/macros/")) { return XmlType.MACRO }
+    if (isChild(child, "../idprovider/")) { return XmlType.ID_PROVIDER }
+    if (isChild(child, "../site/")) { return XmlType.SITE }
+    return null
+  }
```

Comment:
Don't use `it` as a variable name, since it is a keyword. Use a name that gives more meaning.

Let us not use `null` at all. Lets return `Option<Xmltype>`, using [Option](https://arrow-kt.io/docs/apidocs/arrow-core-data/arrow.core/-option/) from arrow.

What you should have in the calling function then is a `Sequence<Option<XmlType>>`. In scala you can do a `flatten` to get a `Sequence<Xmltype>` from it. But you might need to use another technique here.

I think this would be nicer if you return  `when`-expression (instead of having `if`-statements that has returns).

```suggestion
  private fun getXmlType(it: String): Option<XmlType> {
    val child: Path = Paths.get(it).toAbsolutePath()

    return when {
      isChild(child, "../site/content-types/") -> Some(XmlType.CONTENT_TYPE)
      isChild(child, "../site/layouts/") -> Some(XmlType.LAYOUT)
      isChild(child, "../site/mixins/") -> Some(XmlType.MIXIN)
      isChild(child, "../site/parts/") -> Some(XmlType.PART)
      isChild(child, "../site/pages/") -> Some(XmlType.PAGE)
      isChild(child, "../site/macros/") -> Some(XmlType.MACRO)
      isChild(child, "../idprovider/") -> Some(XmlType.ID_PROVIDER)
      isChild(child, "../site/") -> Some(XmlType.SITE)
      else -> None
    }
  }
```

Reply:
Did a refactoring here using a list of Pair(), and a loop. Returning an Either.